### ISSUE TITLE
Suboptimal IR after new InstCombine transformation

### ISSUE THREAD

**Issue by aymanmusa** [llvm/llvm-project#54750]
New transformation was introduced in [commit](https://github.com/llvm/llvm-project/commit/1a60ae02c65d26981017f59bc5918d3c2e363bfd).

Attached are:
1. reproducer - reprod.ll
2. result without the transformation - reprod-after-orig.ll
3. result with the transformation - reprod-after.ll

You can easily see that the IR we get without the transformation is very much simpler (simple select), while it gets expanded to long sequence after involving the new transformation.

To run reproduce:
> opt -S -O3 reprod.ll

[reprod-after.zip](https://github.com/llvm/llvm-project/files/8416094/reprod-after.zip)

**Comment by LebedevRI** [comment]
Can you post the reproducer via a link to godbolt?

**Comment by aymanmusa** [comment]
> 

https://godbolt.org/z/5xn1z7rsG
Sorry, but it's been years since I opened the last ticket in LLVM.
Is this ok?

**Comment by LebedevRI** [comment]
There's a number of missing folds.

The first one seems to be:
```
  %and.lobit.neg = ashr i32 %and, 31
  %1 = and i32 %and.lobit.neg, 1
  %.not = icmp eq i32 %1, 0
```
This just checks that the `%and` is non-negative (`s>= 0`): https://alive2.llvm.org/ce/z/t2xBnv

That improves things: https://godbolt.org/z/4YnxKzvz8

Which shows a missing canonicalization of `icmp sgt i4 %and, -1` to `~(icmp slt i4 %and, 0)`

After which we arrive at the optimal ir: https://godbolt.org/z/xKEqKKhK3

**Comment by rotateright** [comment]
Thanks for the report and diagnosis. I am away from my dev machine until next week, but I can work on this as soon as I’m back. If it is critical, please revert the patch that exposed the regression, and we can reapply after the other folds are added.

**Comment by rotateright** [comment]
I added an explicit fold for ashr -> lshr (with one-use it already reduces using demanded bits), and that gets us to the same place as the suggested icmp fold.

**Comment by rotateright** [comment]
I'm still catching up on mails/bugs, but IIUC, 14.0.1 ends today, so I filed another bug to see if we can get the 1st fix in. Note that it does appear to be enough to produce the optimal x86 asm on its own (even though the IR is not optimal yet).

**Comment by rotateright** [comment]
The 2nd (final) change for this bug has been in a trunk for almost a week now, so let's see if we can get that into 14.0.2 and close this bug:
/cherry-pick 2c2568f39ec641aa8f1dcc01

**Comment by tstellar** [comment]
Merged: 9a3e81e1f91f
